Export `import_account_by_id` to WebClient

partylikeits1983 opened the pull request:
This PR adds the `import_account_by_id` functionality that was added in PR #733 to the WebClient. This resolves issue #855.

This PR makes it possible to import a public account on the WebClient:
```ts
const account = await client.importAccountById("0xb584c3769ab90b000004c780363668");
```
This function only works if the Account is public. Obviously this would only be used for public accounts that do not require authentication when submitting state changing transactions, i.e. a counter contract.

I added a test that deploys a public account, then deletes the store, then reimports it via the `importAccountById` method.

To run the tests:

1) Run the node locally by cloning the node repository, then run the following:
```
miden-node bundled bootstrap \
  --data-directory data \
  --accounts-directory accounts \
  --config genesis.toml
```
Start the node:
```
miden-node bundled start \
  --data-directory data \
  --rpc.url http://0.0.0.0:57291
```

2) Checkout this branch and run
```
cd crates/web-client
npm i
npm test -- --grep "import public account by id" --invert=false
```

If all looks good, since this is a non breaking change, would it be possible to merge it to main and publish an updated version of the WebClient-SDK?

igamigo commented on crates/web-client/src/import.rs:
```diff
@@ -69,6 +73,33 @@ impl WebClient {
     }
 
+    #[wasm_bindgen(js_name = "importAccountById")]
+    pub async fn import_account_by_id(&mut self, account_id: String) -> Result<Account, JsValue> {
```
There is an `AccountId` web struct already, so we could use that instead, right? `miden_client_web::models::account_id`. I see that some other functions use that instead of strings for account IDs. And then you can convert into the native type by doing `account_id.into()`.

partylikeits1983 replied:
Good idea. I changed to use this.

igamigo commented on crates/web-client/src/import.rs:
```diff
@@ -69,6 +73,33 @@ impl WebClient {
+            let inner_acct = record.account().clone();
+            Ok(Account::from(inner_acct))
+        } else {
+            Err(JsValue::from_str("failed to parse account record"))
```
nit: Maybe this is more precise
```suggestion
            Err(JsValue::from_str("account importing did not fail but the account was not found in the store"))
```

partylikeits1983 replied:
I simplified the exported `import_account_by_id` function, it doesn't need these messages anymore.

igamigo commented on crates/web-client/src/import.rs:
```diff
@@ -69,6 +73,33 @@ impl WebClient {
+            .await
+            .map_err(|err| js_error_with_context(err, "failed to import public account"))?;
+
+        let record_opt = client
+            .get_account(account_id)
+            .await
+            .map_err(|err| js_error_with_context(err, "failed to retrieve account"))?;
+
+        if let Some(record) = record_opt {
+            let inner_acct = record.account().clone();
+            Ok(Account::from(inner_acct))
+        } else {
+            Err(JsValue::from_str("failed to parse account record"))
+        }
```
Is there any benefit to returning the imported account instead of leaving it up to the user? I don't mind either way but maybe we want to return the account from the `rust-client` function directly so that this is simpler as well.

partylikeits1983 replied:
Yes its probably simplest to do:

```js
await client.importAccountById(accountId);
const account = await client.getAccount(accountId);
```

Will change to this.

igamigo reviewed (commented):
Looking good overall, leaving some comments for now.

partylikeits1983 commented:
Refactored so that the `client.importAccountById(accountId)` takes an `AccountId` type. Also this function no longer returns the imported account. To get the imported account the user needs to call, `client.getAccount(accountId)`

The usage looks like this:
```ts
  await client.importAccountById(accountId);
  const account = await client.getAccount(accountId);
```
where `accountId` is an `AccountId` type.

Updated the test to match.

igamigo reviewed (approved):
LGTM! When CI passes we can merge.

But should this branch merge to `next` instead?  If we want to use this in the context of the testnet, it's fine that it goes to `main` but that means we'd have to make a new release (cc @dagarcia7 @bobbinth)

partylikeits1983 commented on crates/web-client/src/import.rs:
```diff
@@ -60,15 +69,33 @@ impl WebClient {
+            .get_mut_inner()
+            .ok_or_else(|| JsValue::from_str("Client not initialized"))?;
+
+        let native_id: NativeAccountId = NativeAccountId::from(account_id);
```
Fixed.

partylikeits1983 commented:
> But should this branch merge to next instead? If we want to use this in the context of the testnet, it's fine that it goes to main but that means we'd have to make a new release

Its ok if this goes to `next` but its a non breaking change since its just adding one function to the WebClient API. If possible, I'd prefer if we could push to main so that there is a stable version of the WebClient that has this feature as opposed to using the `next` version of the WebClient, which at some point may not be backwards compatible with `0.8`. This would make building the counter contract frontend easier. Either way is ok.

partylikeits1983 commented:
@dagarcia7 @igamigo 

Two questions:
1) I was getting this same error as seen in CI tests locally, but when I ran the miden node on port `57291` like this:
```sh
miden-node bundled start \
  --data-directory data \
  --rpc.url http://0.0.0.0:57291
```
All the tests passed. 

2) Does this need a change log?
